File: src/main/java/edu/kit/iti/formal/stvs/model/expressions/ValueBool.java
package edu.kit.iti.formal.stvs.model.expressions;

/**
 * runtime-representation for boolean values of {@link Expression}s.
 * <p>
 * This is a singleton with two instances, TRUE and FALSE, since there
 * is no state to the values.
 *
 * @author Philipp
 */
public class ValueBool implements Value {

  public static final ValueBool TRUE = new ValueBool(true);
  public static final ValueBool FALSE = new ValueBool(false);

  public static ValueBool of(boolean bool) {
    return bool ? TRUE : FALSE;
  }

  private final boolean value;

  private ValueBool(boolean value) {
    this.value = value;
  }

  @Override
  public <R> R match(
      ValueIntegerHandler<R> matchInt,
      ValueBooleanHandler<R> matchBoolean,
      ValueEnumHandler<R> matchEnum) {
    return matchBoolean.handle(value);
  }

  public boolean getValue() {
    return value;
  }

  @Override
  public Type getType() {
    return TypeBool.BOOL;
  }

  @Override
  public String getValueString() {
    if (getValue()) {
      return "TRUE";
    } else {
      return "FALSE";
    }
  }

  public String toString() {
    return "ValueBool(" + value + ")";
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof ValueBool)) {
      return false;
    }

    ValueBool valueBool = (ValueBool) obj;

    return value == valueBool.value;

  }

  @Override
  public int hashCode() {
    return (value ? 1 : 0);
  }
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/model/expressions/ValueBool.kt
package edu.kit.iti.formal.stvs.model.expressions

/**
 * Runtime-representation for boolean values of [Expression]s.
 * This is a singleton with two instances, TRUE and FALSE, since there is no state to the values.
 *
 * @author Philipp
 */
class ValueBool private constructor(val value: Boolean) : Value {
    override fun <R> match(
        matchInt: ValueIntegerHandler<R>, matchBoolean: ValueBooleanHandler<R>,
        matchEnum: ValueEnumHandler<R>
    ): R? {
        return matchBoolean.handle(value)
    }

    override val type: Type
        get() = TypeBool.Companion.BOOL

    override val valueString: String
        get() = if (value) {
            "TRUE"
        } else {
            "FALSE"
        }

    override fun toString(): String {
        return "ValueBool($value)"
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj) {
            return true
        }
        if (obj !is ValueBool) {
            return false
        }

        return value == obj.value
    }

    override fun hashCode(): Int {
        return (if (value) 1 else 0)
    }

    companion object {
        @JvmField
        val TRUE: ValueBool = ValueBool(true)
        @JvmField
        val FALSE: ValueBool = ValueBool(false)

        @JvmStatic
        fun of(bool: Boolean): ValueBool {
            return if (bool) TRUE else FALSE
        }
    }
}
